src/Server/TCPHandler.cpp

```diff
@@ -873,17 +892,75 @@ void TCPHandler::receiveQuery()
         client_info.initial_query_id = client_info.current_query_id;
         client_info.initial_address = client_info.current_address;
     }
-    else
-    {
-        query_context->setInitialRowPolicy();
-    }
 
     /// Per query settings are also passed via TCP.
     /// We need to check them before applying due to they can violate the settings constraints.
     auto settings_format = (client_revision >= DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS) ? SettingsWriteFormat::STRINGS_WITH_FLAGS
                                                                                                       : SettingsWriteFormat::BINARY;
     Settings passed_settings;
     passed_settings.read(*in, settings_format);
+
+    /// Interserver secret.
+    bool interserver_secret = client_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET;
+    std::string received_cluster;
+    std::string received_hash;
+    if (interserver_secret)
+    {
+        readStringBinary(received_cluster, *in);
```

Comment:
Maybe it's better to pass the cluster's name with the hello package.

Reply:
Great idea!

Reply:
Although it can't be done in the Hello package (since it can't be expanded), requires separate package, say `Cluster` (cluster name, salt)

Reply:
We may have the extended hello package. I mean if the client wants to send some extra data (cluster's name) or receive some extra data (salt) it will send the extended hello package instead of the standard one. If the server has an older version of course it won't able to manage that extended hello package, but an older version is not able to manage interserver secret anyway.

Reply:
Ok, added `Cluster` packet type for now (I don't see any pros of adding some super packet that will encapsulate information for lots of places, better to add more packets I would say)

Reply:
Adding a new packet which will be always sent after the hello packet can affect the performance. Though I don't know if we have to worry about the impact.

Reply:
I've talked about the problem with Alexey Milovidov. So there are two solutions:
1) send an extra packet just as you're doing. The advantage of the solution is that it's easier to implement.
2) send an extended hello packet first (with another packet type), then if the client receives an exception it checks the error code and if the error code is UNEXPECTED_PACKET_FROM_CLIENT the client sends the standard hello packet. The advantage of this solution is that the number of packets will be increased only if the server and the client have different versions.

Reply:
>Adding a new packet which will be always sent after the hello packet can affect the performance. Though I don't know if we have to worry about the impact.

Doubt that this will have huge impact, since:
- they will be sent only for distributed queries with interserver secret enabled (http over TLS has a huge overhead too)
- and besides clickhouse-server does not built for running tons of queries per second, so one extra package does not looks like a huge problem to me